solr/core/src/java/org/apache/solr/update/processor/NestedUpdateProcessorFactory.java

```diff
@@ -0,0 +1,120 @@
+    if(!(storeParent || storePath)) {
+      return next;
+    }
+    return new NestedUpdateProcessor(req, rsp, shouldStoreDocParent(req.getSchema()), shouldStoreDocPath(req.getSchema()), next);
+  }
+
+  private static boolean shouldStoreDocParent(IndexSchema schema) {
+    return schema.getFields().containsKey(IndexSchema.PARENT_FIELD_NAME);
+  }
+
+  private static boolean shouldStoreDocPath(IndexSchema schema) {
+    return schema.getFields().containsKey(IndexSchema.PATH_FIELD_NAME);
+  }
+}
+
+class NestedUpdateProcessor extends UpdateRequestProcessor {
+  private static final String PATH_SEP_CHAR = "/";
+  private boolean storePath;
+  private boolean storeParent;
+  private String uniqueKeyFieldName;
+
+
+  protected NestedUpdateProcessor(SolrQueryRequest req, SolrQueryResponse rsp, boolean storeParent, boolean storePath, UpdateRequestProcessor next) {
+    super(next);
+    this.storeParent = storeParent;
+    this.storePath = storePath;
+    this.uniqueKeyFieldName = req.getSchema().getUniqueKeyField().getName();
+  }
+
+  @Override
+  public void processAdd(AddUpdateCommand cmd) throws IOException {
+    SolrInputDocument doc = cmd.getSolrInputDocument();
+    processDocChildren(doc, null);
+    super.processAdd(cmd);
+  }
+
+  private void processDocChildren(SolrInputDocument doc, String fullPath) {
+    int childNum = 0;
+    for(SolrInputField field: doc.values()) {
+      for(Object val: field) {
+        if(!(val instanceof SolrInputDocument)) {
+          // either all collection items are child docs or none are.
+          break;
+        }
+        final String fieldName = field.getName();
+
+        if(fieldName.contains(PATH_SEP_CHAR)) {
+          throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Field name: '" + fieldName
+              + "' contains: '" + PATH_SEP_CHAR + "' , which is reserved for the nested URP");
+        }
+        final String jointPath = fullPath == null ? fieldName: fullPath + PATH_SEP_CHAR + fieldName;
+        SolrInputDocument cDoc = (SolrInputDocument) val;
+        if(!cDoc.containsKey(uniqueKeyFieldName)) {
+          String parentDocId = doc.getField(uniqueKeyFieldName).getFirstValue().toString();
+          cDoc.setField(uniqueKeyFieldName, generateChildUniqueId(parentDocId, fieldName, childNum));
+        }
+        processChildDoc((SolrInputDocument) val, doc, jointPath);
+        ++childNum;
+      }
+    }
+  }
+
+  private void processChildDoc(SolrInputDocument sdoc, SolrInputDocument parent, String fullPath) {
+    if(storePath) {
+      setPathField(sdoc, fullPath);
+    }
+    if (storeParent) {
+      setParentKey(sdoc, parent);
+    }
+    processDocChildren(sdoc, fullPath);
+  }
+
+  private String generateChildUniqueId(String parentId, String childKey, int childNum) {
+    // combines parentId with the child's key and childNum. e.g. "10/footnote/1"
+    return String.join(PATH_SEP_CHAR, parentId, childKey, Integer.toString(childNum));
```

Comment:
Glad you fixed the comment.
Another dubious use of of String.join
What did you think of my suggestion to use some other separator char (not PATH_SEP_CHAR) over here.  My first suggestion was a pound symbol.  Though it wouldn't show when in a URL... maybe a comma.  But we could stay with a '/'; I just thought it might be nice to make a distinction between the separator between parent/child and the separator to append the child sequence/num.
We ought to surface this to the JIRA as it's likely to get input.